src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java

```diff
@@ -328,7 +328,23 @@ boolean open() {
                 throw new InternalError("Native library " + name + " has been loaded");
             }
 
-            return load(this, name, isBuiltin, loadLibraryOnlyIfPresent);
+            return load(this, name, isBuiltin, throwExceptionIfFail());
+        }
+
+        @SuppressWarnings("removal")
+        private boolean throwExceptionIfFail() {
+            if (loadLibraryOnlyIfPresent) return true;
+
+            boolean fileExists = AccessController.doPrivileged(new PrivilegedAction<>() {
+                public Boolean run() {
+                    File file = new File(name);
+                    return file.exists();
+                }
+            });
+
+            // If the file exists but fails to load, UnsatisfiedLinkException thrown by the VM
+            // will include the error message from dlopen to provide diagnostic information
+            return fileExists;
```

Comment:
Can drop the `fileExists` variable and simply `return AccessController.doPrivileged(....);`

Otherwise, looks good.